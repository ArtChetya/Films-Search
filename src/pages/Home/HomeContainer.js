import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { queryToParams } from 'utils';
import { params as paramsAction } from 'features/searchParams';
import { fetchFilms } from 'features/films';
import { serverSideRenderedFlag } from 'features/serverSideRendered';
import { Home } from '.';

export const HomeContainer = ({
    match,
    history,
    sortBy,
    setSortBy,
    fetchFilmsIfNeeded,
    setParams,
    isFilmsLoading
}) => {
    const sortByOptionsList = [
        { id: 'release_date', label: 'release date' },
        { id: 'vote_average', label: 'rating' }
    ];
    const [sortByOptions] = useState(sortByOptionsList);

    const onSearch = (search, searchBy) => {
        const isByTitleSearch = searchBy === 'title';

        const searchVal = isByTitleSearch ? search : search.split(' ')[0];

        history.push(
            encodeURI(
                `/search/search=${searchVal} searchBy=${searchBy} sortBy=${sortBy}`
            )
        );
    };

    const onSort = id => {
        if (id !== sortBy) {
            setSortBy(id);
            fetchFilms();
        }
    };

    useEffect(() => {
        const queryParams = queryToParams(match.params.query);

        const defaultParams = {
            search: '',
            searchBy: 'title',
            sortBy: 'release_date',
            sortOrder: 'desc',
            limit: 50
        };

        setParams({
            ...defaultParams,
            ...queryParams
        });

        fetchFilmsIfNeeded();
    }, [setParams, fetchFilmsIfNeeded, match.params.query]);

    return (
        <Home
            sortByOptions={sortByOptions}
            sortById={sortBy}
            setSortById={onSort}
            onSearch={onSearch}
            isLoading={isFilmsLoading}
        />
    );
};

HomeContainer.onInit = (store, match) => {
    const queryParams = queryToParams(decodeURI(match.params.query));
    const defaultParams = {
        search: '',
        searchBy: 'title',
        sortBy: 'release_date',
        sortOrder: 'desc',
        limit: 50
    };

    store.dispatch(
        paramsAction({
            ...defaultParams,
            ...queryParams
        })
    );

    store.dispatch(serverSideRenderedFlag(true));

    return store.dispatch(fetchFilms());
};

HomeContainer.propTypes = {
    match: PropTypes.object.isRequired,
    history: PropTypes.object.isRequired,
    sortBy: PropTypes.string.isRequired,
    setSortBy: PropTypes.func.isRequired,
    fetchFilmsIfNeeded: PropTypes.func.isRequired,
    setParams: PropTypes.func.isRequired,
    isFilmsLoading: PropTypes.bool.isRequired
};
