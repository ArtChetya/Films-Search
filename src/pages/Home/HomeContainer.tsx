import { fetchFilms } from 'features/films';
import {
    ISearchParamsState,
    params as paramsAction
} from 'features/searchParams';
import { serverSideRenderedFlag } from 'features/serverSideRendered';
import React, { FunctionComponent, useEffect, useState } from 'react';
import { RouteComponentProps } from 'react-router';
import { queryToParams } from 'utils';
import { Home } from '.';
import { IHomePageDispatchProps, IHomePageStateProps } from './HomeConnector';
import { IOption } from './types';

interface IParams {
    query: string;
}

interface IHomeContainerProps
    extends RouteComponentProps<IParams>,
        IHomePageStateProps,
        IHomePageDispatchProps {}

export const HomeContainer: FunctionComponent<IHomeContainerProps> & {
    onInit: (store: any, match: any) => any;
} = ({
    match,
    history,
    sortBy,
    setSortBy,
    fetchFilmsIfNeeded,
    setParams,
    isFilmsLoading
}) => {
    const sortByOptionsList: IOption[] = [
        { id: 'release_date', label: 'release date' },
        { id: 'vote_average', label: 'rating' }
    ];
    const [sortByOptions] = useState<IOption[]>(sortByOptionsList);

    const onSearch = (search: string, searchBy: string) => {
        const isByTitleSearch = searchBy === 'title';

        const searchVal = isByTitleSearch ? search : search.split(' ')[0];

        history.push(
            encodeURI(
                `/search/search=${searchVal} searchBy=${searchBy} sortBy=${sortBy}`
            )
        );
    };

    const onSort = (id: string) => {
        if (id !== sortBy) {
            setSortBy(id);
            fetchFilmsIfNeeded();
        }
    };

    useEffect(() => {
        const queryParams = queryToParams(match.params.query);

        const defaultParams = {
            search: '',
            searchBy: 'title',
            sortBy: 'release_date',
            sortOrder: 'desc',
            limit: 50
        };

        setParams({
            ...defaultParams,
            ...queryParams
        });

        fetchFilmsIfNeeded();
    }, [setParams, fetchFilmsIfNeeded, match.params.query]);

    return (
        <Home
            sortByOptions={sortByOptions}
            sortById={sortBy}
            setSortById={onSort}
            onSearch={onSearch}
            isLoading={isFilmsLoading}
        />
    );
};

HomeContainer.onInit = (store, match) => {
    const queryParams = queryToParams(decodeURI(match.params.query));
    const defaultParams: ISearchParamsState = {
        search: '',
        searchBy: 'title',
        sortBy: 'release_date',
        sortOrder: 'desc',
        limit: 50
    };

    store.dispatch(
        paramsAction({
            ...defaultParams,
            ...queryParams
        })
    );

    store.dispatch(serverSideRenderedFlag(true));

    return store.dispatch(fetchFilms());
};
