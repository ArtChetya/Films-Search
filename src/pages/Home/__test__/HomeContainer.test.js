import { shallow } from 'enzyme/build';
import React from 'react';
import { HomeContainer } from '../HomeContainer';

jest.mock('../../../services/httpService');

describe('HomeContainer component', () => {
    let sortBy;
    let setSortBy;
    let fetchFilms;
    let params;
    let setParams;
    let isFilmsLoading;

    beforeAll(() => {
        sortBy = 'release_date';

        setSortBy = jest.fn();

        fetchFilms = jest.fn();

        params = {
            search: undefined,
            searchBy: 'title',
            sortBy: 'release_date',
            sortOrder: 'desc',
            limit: 50
        };

        setParams = jest.fn();

        isFilmsLoading = false;
    });

    it('should be rendered correctly', () => {
        const component = shallow(
            <HomeContainer
                sortBy={sortBy}
                setSortBy={setSortBy}
                fetchFilms={fetchFilms}
                params={params}
                setParams={setParams}
                isFilmsLoading={isFilmsLoading}
            />
        );

        expect(component).toMatchSnapshot();
    });
});
