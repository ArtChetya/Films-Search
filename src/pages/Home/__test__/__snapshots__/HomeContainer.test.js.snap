// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeContainer component should be rendered correctly 1`] = `
<Home
  films={Array []}
  isLoading={false}
  onSearch={[Function]}
  search=""
  searchById="title"
  searchByOptions={
    Array [
      Object {
        "id": "title",
        "label": "Title",
      },
      Object {
        "id": "genres",
        "label": "Genre",
      },
    ]
  }
  setSearch={[Function]}
  setSearchById={[Function]}
  setSortById={[Function]}
  sortById="release_date"
  sortByOptions={
    Array [
      Object {
        "id": "release_date",
        "label": "release date",
      },
      Object {
        "id": "vote_average",
        "label": "rating",
      },
    ]
  }
/>
`;
